app/gui2/src/providers/widgetRegistry.ts

```diff
@@ -69,24 +109,75 @@ export enum Score {
   Perfect,
 }
 
-export interface WidgetProps {
-  input: WidgetInput
+export interface WidgetProps<T> {
+  input: T
   config: WidgetConfiguration | undefined
   nesting: number
 }
 
-export interface WidgetDefinition {
+/**
+ * Create vue props definition for a widget component. This cannot be done automatically by using
+ * typed `defineProps`, because vue compiler is not able to resolve conditional types. As a
+ * workaround, the runtime prop information is specified manually, and the complex conditional type
+ * is still provided through `PropType`.
+ *
+ */
+export function widgetProps<Def extends WidgetDefinition<any>>(_def: Def) {
+  return {
+    input: {
+      type: Object as PropType<Def extends WidgetDefinition<infer T> ? T : WidgetInput>,
+      required: true,
+    },
+    config: { type: Object as PropType<WidgetConfiguration | undefined>, required: false },
+    nesting: { type: Number, required: true },
+  } as const
+}
+
+type InputMatchFn<T extends WidgetInput> = (input: WidgetInput) => input is T
+
+type InputTy<M> = M extends InputMatchFn<infer T>
```

Comment:
`M extends InputMatchFn<infer T> | InputMatchFn<infer T>[]` might work. but this may not be necessary (see below)

Reply:
I didn't know that `infer` can be used multiple times on the same variable within a signle conditional. Cool!